server/server.go

```diff
@@ -1414,20 +1414,31 @@ func (s *BgpServer) DeleteBmp(c *config.BmpServerConfig) error {
 
 func (s *BgpServer) Shutdown() {
 	s.mgmtOperation(func() error {
-		s.shutdown = true
-		stateOp := AdminStateOperation{ADMIN_STATE_DOWN, nil}
+		s.shutdownWG = new(sync.WaitGroup)
+		s.shutdownWG.Add(1)
+		stateOp := AdminStateOperation{
+			State:         ADMIN_STATE_DOWN,
+			Communication: nil,
+		}
 		for _, p := range s.neighborMap {
 			p.fsm.adminStateCh <- stateOp
 		}
-		// the main goroutine waits for peers' goroutines to stop but if no peer is configured, needs to die immediately.
```

Comment:
We can just do a `s.shutdownWG.Wait()` here since there is already a `Stop` that does not wait on these idle neighbors. This also eliminates the need for another method to run after a Shutdown. 

Imagine the `Shutdown` call being the "graceful" handling of `server.Stop` + wait for neighbors to be idle.

Reply:
Yeah, I know `Shutdown()` and `Stop()` are very similar, and I tried to combine shutting down and waiting features into a single `Shutdown()` function. But as I mentioned as https://github.com/osrg/gobgp/pull/1757#discussion_r197689335, if we do `s.shutdownWG.Wait()` within `s.mgmtOperation()` wrapper function, it causes the deadlock of the main goroutine.

So, how about putting `s.shutdownWG.Wait()` in `Shutdown()` and the outside of `s.mgmtOperation()` wrapper like;
```go
func (s *BgpServer) Shutdown() {
	s.mgmtOperation(func() error {
		// ...(snip)...
	}, false)

	// Waits for all goroutines per peer to stop.
	// Note: This should not be wrapped with s.mgmtOperation() in order to
	// avoid the deadlock in the main goroutine of BgpServer.
	if s.shutdownWG != nil {
		s.shutdownWG.Wait()
		s.shutdownWG = nil
	}
}
```

Reply:
Perfect. That makes this a lot easier to read through and keeps the waiting within the Shutdown function. 

Looks great!

Reply:
Thunks for your confirmation!